### PR TITLE
Initial implementation of LX/LE file plugin ##bin

### LINKED ISSUE #314: Support for loading LE files (OS/2) executables
Support for LE files/

OS/2

32-bit Linear Executable
    Introduced with OS/2 2.0, these can be identified by the "LX" in ASCII. These can only be run by OS/2 2.0 and higher.[citation needed] They are also used by some DOS extenders.
Mixed 16/32-bit Linear Executable
    Introduced with OS/2 2.0, these can be identified by the "LE" in ASCII. This format is not used for OS/2 applications anymore, but instead for VxD drivers under Windows 3.x and Windows 9x, and by some DOS extenders. 

Here are a few examples:
[BIGRAM.ZIP](https://github.com/radare/radare2/files/2643966/BIGRAM.ZIP)
[CP.ZIP](https://github.com/radare/radare2/files/2643967/CP.ZIP)
[CURSOR.ZIP](https://github.com/radare/radare2/files/2643969/CURSOR.ZIP)

### PR DESCRIPTION
**Pull request by GustavoLCR**
Fix #314

### CODE REVIEW

**Comment by XVilka** [comment]
Could you please add the test too?

**Comment by radare** [comment]
doesnt looks like green at all

Screenshot 2019-12-28 at 12 58 46

**Comment by radare** [comment]
can you fix the test to make it green?

File: libr/bin/p/bin_le.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	rbin->cb_printf("InitialEip: 0x%04x\n", h->eip);
+	rbin->cb_printf("InitialStackObj: %u\n", h->stackobj);
+	rbin->cb_printf("InitialEsp: 0x%04x\n", h->esp);
+	rbin->cb_printf("Page Size: 0x%04x\n", h->pagesize);
+	if (bin->is_le) {
+		rbin->cb_printf("Last Page Size: 0x%04x\n", h->pageshift);
+	} else {
+		rbin->cb_printf("Page Shift: 0x%04x\n", h->pageshift);
+	}
+	rbin->cb_printf("Fixup Size: 0x%04x\n", h->fixupsize);
```
Comment (radare) [inline libr/bin/p/bin_le.c]: add space before every ( ini here

File: libr/bin/p/bin_le.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	return false;
+}
+
+static void destroy(RBinFile *bf) {
+	r_bin_le_free (bf->o->bin_obj);
+}
+
+static void header(RBinFile *bf) {
+	RBin *rbin = bf->rbin;
+	r_bin_le_obj_t *bin = bf->o->bin_obj;
```
Comment (radare) [inline libr/bin/p/bin_le.c]: maybe do some nullchks and do PrintfCallback p = rbin->cb_printf to make further calls non indirect

File: libr/bin/p/bin_le.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	if (length <= 0x3d) {
+		return false;
+	}
+	ut16 idx = r_buf_read_le16_at (b, 0x3c);
+	if ((ut64)idx + 26 < length) {
+		ut8 buf[2];
+		r_buf_read_at (b, 0, buf, sizeof (buf));
+		if (!memcmp (buf, "LX", 2) || !memcmp (buf, "LE", 2)) {
+			return true;
+		} else if (!memcmp (buf, "MZ", 2)) {
```
Comment (radare) [inline libr/bin/p/bin_le.c]: no need to else after return

File: libr/bin/p/bin_le.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
+/* radare - LGPL - Copyright 2019 - GustavoLCR */
+
+#include <r_bin.h>
+#include "../format/le/le.h"
+
+static bool check_buffer(RBuffer *b) {
+	ut64 length = r_buf_size (b);
+	if (length <= 0x3d) {
```
Comment (radare) [inline libr/bin/p/bin_le.c]: check if lentgh is > 1 at least

File: libr/bin/format/le/le.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	bool is_le; /* Used for differences between LE and LX */
+	char *filename;
+	const char *type;
+	const char *cpu;
+	const char *os;
+	const char *arch;
+	ut32 headerOff; /* File offset to start of LE/LX header */
+	LE_object_entry *objtbl;
+	void *buf; /* Pointer to RBuffer of file */
+} r_bin_le_obj_t;
```
Comment (radare) [inline libr/bin/format/le/le.h]: we use to have camelcased typedefs. use `RBinLEObj` instead

File: libr/bin/format/le/le_specs.h [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+	R_PACKED (struct {
+		ut8 flags;   // First bit set if exported, mask with 0xF8 to get parameters count
+		ut32 offset; // This is the offset in the object for the entry point defined at this ordinal number.
+	}) entry_32;
+	R_PACKED (struct {
+		ut8 flags; // First bit set if import by ordinal
+		ut16 import_ord; // This is the index into the Import Module Name Table for this forwarder.
+		ut32 offset; // If import by ordinal, is the ordinal number into the Entry Table of the target module, else is the offset into the Procedure Names Table of the target module.
+	}) forwarder;
+}) LE_entry_bundle_entry;
```
Comment (radare) [inline libr/bin/format/le/le_specs.h]: camelcased? strange nested rpacked usage btw :?

File: libr/bin/p/bin_le.c [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	}
+	ut16 idx = r_buf_read_le16_at (b, 0x3c);
+	if ((ut64)idx + 26 < length) {
+		ut8 buf[2];
+		r_buf_read_at (b, 0, buf, sizeof (buf));
+		if (!memcmp (buf, "LX", 2) || !memcmp (buf, "LE", 2)) {
+			return true;
+		}
+		if (!memcmp (buf, "MZ", 2)) {
+			r_buf_read_at (b, idx, buf, sizeof (buf));
```
Comment (radare) [inline libr/bin/p/bin_le.c]: (void) if you dont chk the return value here

File: libr/bin/format/le/le.c [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+		} else {
+			sec->bits = R_SYS_BITS_16;
+		}
+		sec->is_data = entry->flags & O_RESOURCE || !(sec->perm & R_PERM_X);
+		if (!entry->page_tbl_entries) {
+			r_list_append (l, sec);
+		}
+		int j;
+		ut32 page_size_sum = 0;
+		ut32 next_idx = i < h->objcnt - 1 ? bin->objtbl[i + 1].page_tbl_idx - 1 : UT64_MAX;
```
Comment (pelijah) [inline libr/bin/format/le/le.c]: UT64_MAX?

**Pull request merged**